src/storage/src/source/mod.rs

```diff
@@ -408,9 +411,55 @@ pub(crate) trait SourceReader {
     ///
     /// Note that implementers are required to present messages in strictly ascending\
     /// offset order within each partition.
+    async fn next(
+        &mut self,
+    ) -> Option<Result<SourceMessage<Self::Key, Self::Value>, SourceReaderError>> {
+        // Compatiblity implementation that delegates to the deprecated [Self::get_next_method]
+        // call. Once all source implementations have been transitioned to implement
+        // [SourceReader::next] directly this provided implementation should be removed and the
+        // method should become a required method.
+        loop {
+            match self.get_next_message() {
+                Ok(NextMessage::Ready(msg)) => return Some(Ok(msg)),
+                Err(err) => return Some(Err(err)),
+                // There was a temporary hiccup in getting messages, check again asap.
+                Ok(NextMessage::TransientDelay) => {
+                    tokio::time::sleep(Duration::from_millis(1)).await
+                }
+                // There were no new messages, check again after a delay
+                Ok(NextMessage::Pending) => tokio::time::sleep(Duration::from_secs(1)).await,
+                Ok(NextMessage::Finished) => return None,
+            }
+        }
+    }
+
+    /// Returns the next message available from the source.
+    ///
+    /// Note that implementers are required to present messages in strictly ascending\
+    /// offset order within each partition.
+    ///
+    /// # Deprecated
+    ///
+    /// Source implementation should implement the async [SourceReader::next] method instead.
     fn get_next_message(
         &mut self,
     ) -> Result<NextMessage<Self::Key, Self::Value>, SourceReaderError>;
+
+    /// Returns an adapter that treats the source as a stream.
+    ///
+    /// The stream produces the messages that would be produced by repeated calls to `next`.
+    fn into_stream(
+        mut self,
+    ) -> Pin<Box<dyn Stream<Item = Result<SourceMessage<Self::Key, Self::Value>, SourceReaderError>>>>
+    where
+        Self: Sized + 'static,
+    {
+        Box::pin(async_stream::stream! {
```

Comment:
does this need to be boxed? I think it can just be pinned with `tokio::pin!` when we actually use it later

Reply:
It needs to be boxed for a few reasons. First, the macro is returning a type that contains an `impl Future`, that I can't name in the trait, then you can't use `tokio::pin!` inside a timely closure because timely doesn't pin the closure itself, and finally even if it did pin them you'd have to use `Pin::new_unchecked(..)` because the closure is an `FnMut` which means you won't get an owned value, which is what `tokio::pin!` requires